### PR TITLE
Feature/short circuit calculation python api

### PR DESCRIPTION
**Pull request by mgovers**
* Add short circuit Python API
* Add short circuit validation test cases to the Python tests (but mark `skip`)
* Update the validation data with fixes to the values (see https://github.com/Alliander/pgm-test-generator/pull/5 )

### CODE REVIEW

File: src/power_grid_model/core/power_grid_model.py [hunk 1]
```diff
@@ -309,7 +309,7 @@ def calculate_power_flow(
                 > 0 specify number of parallel threads
             output_component_types: list or set of component types you want to be present in the output dict.
                 By default all component types will be in the output
-            continue_on_batch_error: if the program continues (instead of throwing error) if some scenarios fails
+            continue_on_batch_error: if the program continues (instead of throwing error) if some scenarios fail
```
Comment (mgovers) [inline src/power_grid_model/core/power_grid_model.py:312]: or [...] `scenario fails`

File: tests/data/short_circuit/single_phase_to_ground_b/sc_output_batch.json
Comment (mgovers) [inline tests/data/short_circuit/single_phase_to_ground_b/sc_output_batch.json:1]: here and below: this diff is most likely related to a bug in current main of the test generation. See also https://github.com/Alliander/pgm-test-generator/pull/5 . Please carefully review that PR. If any issues are discovered there, the tests in this repo can be regenerated.

File: tests/unit/utils.py [hunk 2]
```diff
@@ -22,55 +22,114 @@
...
+            params["atol"],
+        ]
+        kwargs = {}
+        if "fail" in params:
+            kwargs["marks"] = pytest.mark.xfail(reason=params["fail"], raises=AssertionError)
+        yield pytest.param(
+            *pytest_param,
+            **kwargs,
+            id=case_id,
+            marks=pytest.mark.skipif(calculation_type == "short_circuit", reason="Short circuit is not yet supported"),
```
Comment (mgovers) [inline tests/unit/utils.py:87]: I commented out this line to test all other changes. Doing so results in the expected error: `power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!`

Example:

```
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_a-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_b-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_c-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/three_phase_abc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
```

Note: I did not implement `validate_input_data` or `validate_batch_data`. Only the prep work for the regular calculation validation tests are implemented here.

File: tests/unit/utils.py [hunk 3]
```diff
@@ -22,55 +22,114 @@
...
+            params["atol"],
+        ]
+        kwargs = {}
+        if "fail" in params:
+            kwargs["marks"] = pytest.mark.xfail(reason=params["fail"], raises=AssertionError)
+        yield pytest.param(
+            *pytest_param,
+            **kwargs,
+            id=case_id,
+            marks=pytest.mark.xfail(calculation_type == "short_circuit", reason="Short circuit is not yet supported"),
```
Comment (mgovers) [inline tests/unit/utils.py:87]: I commented out this line to test all other changes. Doing so results in the expected error: `power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!`

Example:

```
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_a-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_b-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_c-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/three_phase_abc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
```

Note: I did not implement `validate_input_data` or `validate_batch_data`. Only the prep work for the regular calculation validation tests are implemented here.

File: tests/unit/utils.py [hunk 4]
```diff
@@ -22,55 +22,111 @@
...
+            calculation_type,
+            calculation_method,
+            params["rtol"],
+            params["atol"],
+        ]
+        kwargs = {}
+        if "fail" in params:
+            kwargs["marks"] = pytest.mark.xfail(reason=params["fail"], raises=AssertionError)
+        elif calculation_type == "short_circuit":
+            kwargs["marks"] = (pytest.mark.xfail(reason="Short circuit is not yet supported"),)
```
Comment (mgovers) [inline tests/unit/utils.py:84]: I commented out this line to test all other changes. Doing so results in the expected error: power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!

Example:

FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_a-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_b-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/single_phase_to_ground_c-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/three_phase_abc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ab-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_ac-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!
FAILED tests/unit/test_0Z_model_validation.py::test_batch_validation[short_circuit/two_phase_to_ground_bc-asym-iec60909-batch] - power_grid_model.errors.PowerGridError: CalculationType is not implemented for __int64 #2!

Note: I did not implement validate_input_data or validate_batch_data. Only the prep work for the regular calculation validation tests are implemented here.

**Review by TonyXiang8787 — CHANGES_REQUESTED** [review]
Two points:

1. In the generate output arrays, components need to be excluded based on calculation type.
    1. Power flow: exclude all sensor objects and fault objects
    1. State estimation: exclude all fault objects
    1. Short circuit: exclude all sensor objects
2. Since `fault_type` and `fault_phase` are now part of the updatable attribute. We do not need so many test cases anymore. Just make one big test case with all batch scenarios of all fault object, fault phase, fault type, and status of the shunt. Maybe ask @nitbharambe to generate new test case.

**Comment by mgovers** [comment]
> Two points:
> 
>     1. In the generate output arrays, components need to be excluded based on calculation type.
>        
>        1. Power flow: exclude all sensor objects and fault objects
>        2. State estimation: exclude all fault objects
>        3. Short circuit: exclude all sensor objects

Fixed in [7e1f53e](https://github.com/PowerGridModel/power-grid-model/pull/289/commits/7e1f53ee4d3f3a3a2e1374af2f916c1419f8d316). I already did not change the `power_flow` output type exclusion, but now I also generalized that logic to the other calculation types

>     2. Since `fault_type` and `fault_phase` are now part of the updatable attribute. We do not need so many test cases anymore. Just make one big test case with all batch scenarios of all fault object, fault phase, fault type, and status of the shunt. Maybe ask @nitbharambe to generate new test case.

maybe it's a good idea to do the merging of the tests in a separate PR. we're already touching the tests (using the test generator) and i believe keeping those things separate is a good idea in this case. Imo, it's an enhancement to the tests, not to the implementation (which is what this PR does)

**Review by TonyXiang8787 — APPROVED** [review]

**Pull request merged**